Add dependents ("via ..." comments) in export command

zmeir opened the pull request:
Adding dependency trace/parent comments ("via ...") to the export command output.  
This is a similar behavior to the pip compile output.

#### Note to the eager reviewer:
First of all - thanks!  
Secondly, this is still a very rough draft. These are the first lines of code I've ever written in Rust. This is still mostly an educational/fun exercise for myself. If opening a Draft PR is creating too much noise - I apologize and I will close it until it is ready.

## Summary

Resolves #7777

## Test Plan

- [X] manual command execution
- [x] update expected output in tests

charliermarsh commented on crates/uv-resolver/src/lock/requirements_txt.rs:
```diff
@@ -312,6 +314,27 @@ impl<'lock> RequirementsTxtExport<'lock> {
                 marker: reachability.remove(&index).unwrap_or_default(),
+                dependents: graph
+                    .edges_directed(index, Direction::Incoming)
+                    .sorted_by_key(|edge| {
```
What is the purpose of this sort?

zmeir replied:
The purpose is to put on top any dependents that have the root as a direct dependent. This being the current project.

This is before sorting:
<img width="385" alt="image" src="https://github.com/user-attachments/assets/61cfa161-bfc8-4f5d-adb4-cccc6f3d7e6d" />

And this is after:
<img width="385" alt="image" src="https://github.com/user-attachments/assets/b4939efd-ee13-4fe0-a815-08f55005f6fa" />

I will add a secondary sort by name

charliermarsh replied:
Personally I think I'd prefer to sort by name -- it's a little easier to parse visually since it's predictable.

zmeir replied:
Do you mean just by name or first the project and then the rest of the dependents by name?

I tried to be consistent with the output of `uv pip compile`.

charliermarsh replied:
I was proposing sorting _just_ by the dependent name. Doesn't `uv pip compile` just sort by name?

```rust
let mut dependents = graph
    .edges_directed(index, Direction::Incoming)
    .map(|edge| &graph[edge.source()])
    .map(uv_distribution_types::Name::name)
    .collect::<Vec<_>>();
dependents.sort_unstable();
dependents.dedup();
dependents
```

charliermarsh reviewed (commented):
This looks pretty reasonable overall.

zmeir commented:
Thanks for your review @charliermarsh, I think this is just about ready.

I've updated the sorting as discussed [here](https://github.com/astral-sh/uv/pull/12350#discussion_r2017587628). If you would like me to change it please let me know.

I've also added a CLI option: `--no-annotate`, similarly to `uv pip compile`.

eliasecchig commented:
can you consider changing the behavior so to no-annotate by default? 

This is breaking a many frameworks depending on uv

charliermarsh commented:
Sorry, can you expand on that comment? What frameworks are depending on uv, and why are they breaking here? Comments are part of the output format.

eliasecchig commented:
The default output of a specific `uv export` command changed between `uv <= 0.6.10` and `uv >= 0.6.11`, representing a **backwards-incompatible change.**

I am using uv and the export function as part of a public project - see[ this line](https://github.com/GoogleCloudPlatform/agent-starter-pack/blob/main/src/base_template/deployment/cd/staging.yaml#L121C1-L122C1)

```bash
 uv export --no-hashes --no-sources --no-header --no-emit-project --frozen > .requirements.txt
```
**In short:**
*   `uv <= 0.6.10`: Produces **Output X**
*   `uv >= 0.6.11`: Produces **Output Y**

charliermarsh commented:
Right, but the output is still a valid `requirements.txt` file.

zanieb commented:
Sorry, but a change in the output is not considered a breaking change if the _format_ is unchanged. Here, we've just added comments that are compatible with the file format.

eliasecchig commented:
Understood, I will deal with it on my side thanks!
